### ISSUE TITLE
Bug: Scheduled results not persisting across runs

### ISSUE THREAD

**Issue by mdwelsh** [Mobiperf/MobiPerf#42]
I have been running the MobiPerf app and much of the time, when I start it up the "Scheduled measurements" tab under "Results" is empty (well, it contains TWO lines that say that the measurements will appear there). Whatever logic is supposed to be persisting the results across app activations is not working.

**Comment by mdwelsh** [comment]
This was on "V1.0". Will retest with v0.2.

**Comment by mdwelsh** [comment]
This is still a problem in v0.2.

**Comment by huangshu91** [comment]
I just wanted to confirm that others are not getting this problem still.  If I completely stop and restart the app or even reboot my device the results are still persistent for me.

**Comment by mdwelsh** [comment]
This still seems to be a problem. I just pulled my phone out and
looked at the list of results and only see ~30 or so measurements from
the last hour. Where are all of the others?

On Wed, May 30, 2012 at 10:39 AM, huangshu91
reply@reply.github.com
wrote:

**Comment by quietbamboo** [comment]
I just tested and could not tell why the scheduled result log is buggy.
AFAICT, the results show up as expected in my phone with 100% battery under 3G.

Notice that there are a couple of reasons that scheduled experiments will not be run, hence no result in the result tab:
1) If you are using WiFi (WiFi experiments are skipped, same as speedometer)
2) If you battery capacity is below threshold, 75% by default.

Could it be possible that your phone connects to WiFi or your battery goes below 75% so that some scheduled measurements are skipped?

**Comment by mdwelsh** [comment]
Yes, some of them are probably skipped, but I still expect results
older than an hour to show up there. Is there a limit on how many
results are remembered?

On Thu, May 31, 2012 at 1:25 PM, Junxian Huang
reply@reply.github.com
wrote:

**Comment by quietbamboo** [comment]
Yes, at most 128 results. We can change though to whatever we want, or remove the limit if necessary, see the following code from MeasurementScheduler.java

/**
     \* Inserts a string into the console with the latest message on top
     */
    private void insertStringToConsole(ArrayAdapter<String> adapter, String msg) {
        if (msg != null) {
            adapter.insert(msg, 0);
            if (adapter.getCount() > Config.MAX_LIST_ITEMS) {
                adapter.remove(adapter.getItem(adapter.getCount() - 1));
            }
        }
    }

**Comment by mdwelsh** [comment]
I am seeing a lot less than 128 results.
In fact, usually I see zero results, and nothing in the task queue
when I start the app.

The system log is always empty, which was the mechanism we used to use
to debug these issues, so it would be good if you could fix that, it
has been broken for a while.

On Thu, May 31, 2012 at 1:50 PM, Junxian Huang
reply@reply.github.com
wrote:

**Comment by huangshu91** [comment]
What do you mean by the system log is empty?  Is it just missing a few
entries that you expect or completely empty?  The log has never been empty
for me and previously it was filled but due to the color, did not display
correctly.  It displays checkins, measurement runs, etc for me.

On Thu, May 31, 2012 at 5:23 PM, Matt Welsh <
reply@reply.github.com

**Comment by mdwelsh** [comment]
The system log has always been empty for me, I never see anything in there.

On Thu, May 31, 2012 at 2:25 PM, huangshu91
reply@reply.github.com
wrote:

**Comment by quietbamboo** [comment]
I don't know why this does not work for you, but my system log was flooded with at least a hundred logs.
Maybe you can run the most recent version and make sure your phone is connected 3G (not WiFi), and the battery is above threshold.

**Comment by quietbamboo** [comment]
Matt:

To make sure, I just sent an APK file via email asking everyone to give a try. You may install it and see if the issue still exists. Thanks!

**Comment by mdwelsh** [comment]
This still seems to be broken for me. My app only shows results from 8am this morning although it has been running for a while.

**Comment by mdwelsh** [comment]
This still seems to be broken for me. When I start up the app I see no scheduled runs, although I've been running the app for a while now and definitely during times with (a) not WiFi and (b) above battery threshold.

Both the system log and the scheduled tasks seem not to be persisting.

Shuheng, could I ask you to please create a build with a lot of debugging logging in it, send it to me, I will install on my phone and after a couple of days send you the output of 'adb bugreport' which will have those logs?

**Comment by quietbamboo** [comment]
We will try hard to solve this bug, looks like this is the most challenging bug remaining ....

**Comment by huangshu91** [comment]
I sent an apk a few days ago with additional logging.  Have you had a chance to run it and if so could you email me the log?   Regarding the system log I thought we completely resolved that issue already.  I think it would be a good idea to once again make sure we are all on the same page by making an up to date version for us to use.

**Comment by mdwelsh** [comment]
I'm doing some more investigation on this and noticed that this problem seems to exist in the old Speedometer code as well. It is difficult to reproduce but I think it happens when the Speedometer background service is killed due to a low memory issue. (I noticed this morning that the results and system log pane were wiped out after receiving a phone call.)

Looking more deeply at the Speedometer code, it seems that MeasurementScheduler relies on calls to onDestroy() to persist the measurement results and system log as preferences. However, there is no guarantee that this method will be called on a service, and it's possible that things have changed in more recent versions of Android (I'm on jellybean) causing this to happen.

The right way to do this is probably to use something like a sqlite database to store the measurement results and system console and force them to be persisted on every write to the data structure, rather than keeping them in memory and hoping they will be persisted in onDestroy().

Another approach, albeit a little brute force, is to persist the preferences on every update to the data structure.

One other thing I noticed: The scheduler service keeps the measurements it has completed in the in-memory "pendingTasks" data structure, which will also be wiped out if the service is killed. This means that any measurement results that have been completed but not yet uploaded on checkin will be lost.

I think the right approach here is to save the measurement results in a sqlite db and have the UI components query the db for displaying the information. I'll take a look at how hard this is to do.
